packages/shared/src/hooks/useChecklist.ts

```diff
@@ -0,0 +1,59 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { ChecklistStepType } from '../lib/checklist';
+import { Action, ActionType } from '../graphql/actions';
+
+export type UseChecklistProps = {
+  steps: ChecklistStepType[];
+};
+
+export type UseChecklist = {
+  steps: ChecklistStepType[];
+  openStep: string | undefined;
+  onToggleStep: (action: Action) => void;
+  isDone: boolean;
+  activeStep: ActionType | undefined;
+  completedSteps: ChecklistStepType[];
+};
+
+const useChecklist = ({ steps }: UseChecklistProps): UseChecklist => {
+  const activeStep = useMemo(
+    () => steps.find((item) => !item.action.completedAt)?.action.type,
+    [steps],
+  );
+  const [openStep, setOpenStep] = useState<string>(activeStep);
+
+  useEffect(() => {
+    setOpenStep(activeStep);
+  }, [activeStep]);
+
+  const onToggleStep = useCallback((action: Action) => {
+    setOpenStep((currentCheckedStep) => {
+      if (currentCheckedStep === action.type) {
+        return undefined;
+      }
+
+      return action.type;
+    });
+  }, []);
+
+  const isDone = useMemo(() => {
+    return steps.every((item) => !!item.action.completedAt);
+  }, [steps]);
+
+  const completedSteps = useMemo(() => {
+    return steps.filter((step) => !!step.action.completedAt);
+  }, [steps]);
```

Comment:
Do you need useMemo for each individual item if you already memoize the whole return object? (Not 100% sure)
Also tagging @idoshamun from his point of view.

Reply:
We would not need it if we compute everything inside that single `useMemo`, if not `completedSteps` would for example have new reference on each render and then again useMemo at the bottom would recompute on each render making it invalid. The new rule of hooks would actually catch that in this case.

In regards to `useMemo` at the bottom in return, all of our hooks return stable reference so at this point just a consistency.

Reply:
Not sure they would get new reference if steps doesn't change?
I personally think just running useMemo on the return would be enough. *(this just feels like a overkill at this point)

Reply:
I think the one creating new references is the `.filter`.

Reply:
To be honest I would prefer removing return object `useMemo` from all of our hooks since we rarely need reference equality for the whole return value and most often use destructuring for specific properties, which is better pattern of course. Then it is also easier to selectively memoize specific values. 

I am open on removing return object `useMemo` for `useChecklist` and also we can add it as a rule of the thumb for all new hooks.

Reply:
`useMemo` does not control re-renders of the component that is using it. It only optimizes potentially heavy computations of the values it accepts. Having or not having `useMemo` component will re-render if:
1. its state changes (eg. useState)
2. any of the parent props that component (and/or hook) accepts changes

With `useMemo` we by default merely save on computations during those re-renders. Take the above `completedSteps` calculation. Without `useMemo` it would run on every render hence increasing render time. With `useMemo` it will run only if `steps` changes and will use memoized value on every other re-render (skipping actual iteration in `filter`).

Where `useMemo` helps potentially is with re-renders of children components. By default React function components will always re-render if parent component re-renders, which as we said above will happen due to points 1. or 2. Even if we memoize our props children components will re-render if parent component re-renders. This is how React works.

Where we can fruit benefits is if we were to use [memo](https://react.dev/reference/react/memo) on our components. This wrapper will skip rendering components that accept primitive values as props if neither of them change. `memo` also accepts `areEqual` function which can be used to handle non primitive values but much better way of handling this is to just always pass primitive values. How can we do that with arrays and objects who obviously are not primitives and they reference change? We `useMemo` to make them primitive. 

To go back to our `completedSteps` example which is `array` and not primitive value. Without `useMemo` it would have a new reference on each render and our `memo` wrapped component would re-render because it does not know it is actually visually same array. React under the hood uses [Object.is](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) to check that equality. 

With `useMemo` reference will stay the same and `memo` will see it as primitive value. It will only change reference if its dependencies change (in this case `steps`) and only then it will also trigger re-render of our `memo` wrapped children component. This is where we would directly save on re-renders with `useMemo`.

This is why removing `useMemo` from return object would no really do anything unless we were to pass whole return object to some component that uses `memo`. Most of the time we pass specific properties of our return object as props to our children components and this is why I mentioned below statement:

> To be honest I would prefer removing return object useMemo from all of our hooks since we rarely need reference equality for the whole return value and most often use destructuring for specific properties, which is better pattern of course. Then it is also easier to selectively memoize specific values.

In all other cases I like to think about `useMemo` as just a in render optimisation that can potentially (if needed) lead to easier re-render optimisations if we ever need to introduce `memo` on some of our components.

Some additional points:
1. `useMemo` of course helps with `useEffect` which also only knows how to check primitive values, and if `useEffect` calls a state setter then we could say `useMemo` here helped with re-renders eg. if we had `useEffect` that was dependant of our `completedSteps`
2. Should we add `memo` to all of our components, not really, even though it does a lot of work (vs some other frameworks like Svelte and SolidJS) React is plenty fast 99% of the time, only when we see some bottlenecks then we can utilise `memo`
3. We also use some 3party packages and components which do some complex stuff might be wrapped in `memo` by library author so by using `useMemo` we potentially save there
4. There is also a big discussion of `useMemo` overhead vs just relying on V8 runtime optimizations. My rule of the thumb that I sleep better if any kind of iteration or algorithmic function inside components is behind `useMemo`.. but that is a rabbit whole for another day 😆 

I hope I explained it here, would love to talk more about this if needed any time.

Reply:
@idoshamun @rebelchris @sshanzel since we are merging this to epic branch and we are onboard for all the features part of the PR I'll merge this now so we can continue on other squads v5 tasks and I can do another PR to resolve `useMemo` depending on the path we decide to take.

Reply:
Happy to continue the discussion here even though you merged :)
So one thing that you missed is that whenever you define a new return object as json, even if the underlying props are the same it will be referred as a new object and will trigger re-render for those who use this hook. I encountered few years ago when I saw a suggestion to useMemo when returning a json object. Again my assumption might have changed but this is another use case for useMemo

Reply:
> So one thing that you missed is that whenever you define a new return object as json, even if the underlying props are the same it will be referred as a new object and will trigger re-render for those who use this hook.

Yes, this is true, but it would need to be used in specific ways, let's look at specific examples on when would this matter:
1. The whole hook value needs to be passed as dependency so:
```jsx
const checklist = useChecklist()

return <Child checklist={checklist} />
```

If `Child` is component wrapped in `memo` it will now re-render every time if `useChecklist` return object is not using `useMemo` because of course object is not primitive value. This is where return object `useMemo` matters as you said. ✅ 

That said, we mostly have this kind of usages:
```jsx
const checklist = useChecklist()

return <Child {...checklist} />
```

In above example it does not matter if return object is memoized or not it only matters if specific properties are memoized (and they are in the case of `useChecklist`) eg. `completedSteps`, `isDone`...

The above is of course shortcut for something like this:
```jsx
const { completedSteps, isDone } = useChecklist()

return <Child completedSteps={completedSteps} isDone={isDone}  />
```

The same principal applies here, `useMemo` is not needed for return object if we already memoize `completedSteps` and `isDone` separately, since those are actually passed as props separately.

Of course this all comes with a caveat that we have to use `memo` for `Child`, if it is just a function component it will render no matter no matter the props if its parent component(s) re-render.

Your point also stands for `useEffect` where of course we have this:
```jsx
const checklist = useChecklist()

useEffect(() => {
    // do something
}, [checklist])
```

it matters that return object is memoized here since it is directly used as dependency.

Again, we mostly use stuff like this.

```jsx
const { completedSteps } = useChecklist()

useEffect(() => {
    // do something
}, [completedSteps])
```

This is why I think according of how we use our special hooks we do not need to do this double memoization (for return object and separate properties), but leaving it we are not loosing anything drastic and re-renders wise we are not really saving anything since we do not utilise `React.memo` wrapper.

Reply:
I'd love to see a minimal project where we can compare both and see the number of re-renders. You might be right and if so we need to have proper guidelines on that

Reply:
Here is a minimal implementation to see how different examples above re-render https://codesandbox.io/s/re-render-tests-with-usememo-and-memo-p7dg9r (see `index.js` for info on how to use different examples)

Reply:
Thanks for setting it up so quickly. I think we need a bit different example, we need to check what happens when `steps` is updated by the deep value is the same. Whether it will cause re-renders in some cases or not. Anyway, I think it will be better to have a meeting all of us :)
Feel free to schedule.